<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Usage • easyshiny</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Usage">
<meta property="og:description" content="easyshiny">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">easyshiny</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/deploy.html">Deploying your app</a>
    </li>
    <li>
      <a href="../articles/usage.html">Usage</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Versions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/nbisweden/easyshiny">
    <span class="fa fa-github fa-lg fas"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Usage</h1>
                        <h4 data-toc-skip class="author">Roy
Francis</h4>
            
            <h4 data-toc-skip class="date">16-Mar-2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/nbisweden/easyshiny/blob/HEAD/vignettes/usage.Rmd"><code>vignettes/usage.Rmd</code></a></small>
      <div class="hidden name"><code>usage.Rmd</code></div>

    </div>

    
    
<p>This vignette demonstrates the use of <strong>easyshiny</strong>
package to create shiny web app from single-cell transcriptomic
data.</p>
<div class="section level2">
<h2 id="a-small-dataset">A small dataset<a class="anchor" aria-label="anchor" href="#a-small-dataset"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">easyshiny</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span></code></pre></div>
<p>In this example, we will use the built-in data that comes with
Seurat.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small</span></span></code></pre></div>
<pre><code>An object of class Seurat 
230 features across 80 samples within 1 assay 
Active assay: RNA (230 features, 20 variable features)
 2 dimensional reductions calculated: pca, tsne</code></pre>
<p>Since it has been already processed (normalisation, scaling, PCA,
UMAP), we can proceed to building the app.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_conf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_config.html">create_config</a></span><span class="op">(</span><span class="va">pbmc_small</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/make_app.html">make_app</a></span><span class="op">(</span><span class="va">pbmc_small</span>, <span class="va">pbmc_conf</span>, gex.assay<span class="op">=</span><span class="st">"RNA"</span>, gex.slot<span class="op">=</span><span class="st">"data"</span>, shiny.title <span class="op">=</span> <span class="st">"PBMC"</span>, shiny.dir<span class="op">=</span><span class="st">"pbmc"</span><span class="op">)</span></span></code></pre></div>
<p>This creates a directory called <strong>pbmc</strong> which contains
the following files:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="st">"pbmc"</span><span class="op">)</span></span></code></pre></div>
<pre><code>pbmc
├── about.md
├── sc1conf.rds
├── sc1def.rds
├── sc1gene.rds
├── sc1gexpr.h5
├── sc1meta.rds
├── server.R
├── ui.R
└── www
    └── styles.css</code></pre>
<p>To launch the app, run:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">shiny</span><span class="fu">::</span><span class="fu"><a href="https://shiny.rstudio.com/reference/shiny/latest/runApp.html" class="external-link">runApp</a></span><span class="op">(</span><span class="st">"pbmc"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Tip!</strong> Use
<code>Rscript -e 'shiny::runApp("app",launch.browser=F)'</code> if
running from the terminal.</p>
<p><img src="images/preview-pbmc.jpg"></p>
</div>
<div class="section level2">
<h2 id="a-larger-dataset">A larger dataset<a class="anchor" aria-label="anchor" href="#a-larger-dataset"></a>
</h2>
<p>We will now try a larger dataset using the
<strong>SeuratData</strong> package.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://www.satijalab.org/seurat" class="external-link">SeuratData</a></span><span class="op">)</span></span>
<span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/LoadData.html" class="external-link">LoadData</a></span><span class="op">(</span><span class="st">"panc8"</span><span class="op">)</span></span>
<span><span class="va">obj</span></span></code></pre></div>
<pre><code>An object of class Seurat 
34363 features across 14890 samples within 1 assay 
Active assay: RNA (34363 features, 0 variable features)</code></pre>
<p>Run basic processing including normalisation, scaling, PCA and
UMAP.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">obj</span> <span class="op">&lt;-</span> <span class="va">obj</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span>dims<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">panc8_conf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_config.html">create_config</a></span><span class="op">(</span><span class="va">obj</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/make_app.html">make_app</a></span><span class="op">(</span><span class="va">obj</span>, <span class="va">panc8_conf</span>, gex.assay<span class="op">=</span><span class="st">"RNA"</span>, gex.slot<span class="op">=</span><span class="st">"data"</span>, shiny.title <span class="op">=</span> <span class="st">"Panc8"</span>, shiny.dir<span class="op">=</span><span class="st">"panc8"</span><span class="op">)</span></span>
<span><span class="fu">shiny</span><span class="fu">::</span><span class="fu"><a href="https://shiny.rstudio.com/reference/shiny/latest/runApp.html" class="external-link">runApp</a></span><span class="op">(</span><span class="st">"panc8"</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/preview-panc8.jpg"></p>
</div>
<div class="section level2">
<h2 id="multiple-datasets">Multiple datasets<a class="anchor" aria-label="anchor" href="#multiple-datasets"></a>
</h2>
<p>Now we will look at an example using multiple datasets. We will use
the same two datasets used above. Make sure both datasets have been
pre-processed with UMAP.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_conf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_config.html">create_config</a></span><span class="op">(</span><span class="va">pbmc_small</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/make_file.html">make_file</a></span><span class="op">(</span><span class="va">pbmc_small</span>, <span class="va">pbmc_conf</span>, gex.assay<span class="op">=</span><span class="st">"RNA"</span>, gex.slot<span class="op">=</span><span class="st">"data"</span>, shiny.prefix <span class="op">=</span> <span class="st">"sc1"</span>, shiny.dir<span class="op">=</span><span class="st">"app"</span><span class="op">)</span></span>
<span><span class="va">panc8_conf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_config.html">create_config</a></span><span class="op">(</span><span class="va">obj</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/make_file.html">make_file</a></span><span class="op">(</span><span class="va">obj</span>, <span class="va">panc8_conf</span>, gex.assay<span class="op">=</span><span class="st">"RNA"</span>, gex.slot<span class="op">=</span><span class="st">"data"</span>, shiny.prefix <span class="op">=</span> <span class="st">"sc2"</span>, shiny.dir<span class="op">=</span><span class="st">"app"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/make_code_multi.html">make_code_multi</a></span><span class="op">(</span></span>
<span>  shiny.title <span class="op">=</span> <span class="st">"Multi-data"</span>,</span>
<span>  shiny.prefix <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sc1"</span>, <span class="st">"sc2"</span><span class="op">)</span>,</span>
<span>  shiny.headers <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PBMC"</span>, <span class="st">"Panc8"</span><span class="op">)</span>,</span>
<span>  shiny.dir <span class="op">=</span> <span class="st">"app"</span><span class="op">)</span></span></code></pre></div>
<p>This will create a directory named <strong>app</strong> with the
following files corresponding to both datasets:</p>
<pre><code>app
├── about.md
├── sc1conf.rds
├── sc1def.rds
├── sc1gene.rds
├── sc1gexpr.h5
├── sc1meta.rds
├── sc2conf.rds
├── sc2def.rds
├── sc2gene.rds
├── sc2gexpr.h5
├── sc2meta.rds
├── server.R
├── ui.R
└── www
    └── styles.css</code></pre>
<p>Launch the app.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">shiny</span><span class="fu">::</span><span class="fu"><a href="https://shiny.rstudio.com/reference/shiny/latest/runApp.html" class="external-link">runApp</a></span><span class="op">(</span><span class="st">"app"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="images/preview-multiple.jpg" alt=""><p class="caption">Preview showing mutiple datasets.</p>
</div>
</div>
<div class="section level2">
<h2 id="options">Options<a class="anchor" aria-label="anchor" href="#options"></a>
</h2>
<p>We will take a look at some of the customisation options.</p>
<div class="section level3">
<h3 id="selective-tabs">Selective tabs<a class="anchor" aria-label="anchor" href="#selective-tabs"></a>
</h3>
<p>There are 9 different tabs that display different kinds of
information. These are listed below:</p>
<table class="table">
<thead><tr class="header">
<th>Tab</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>civge</td>
<td>Cell info vs gene expression</td>
</tr>
<tr class="even">
<td>civci</td>
<td>Cell info vs cell info</td>
</tr>
<tr class="odd">
<td>gevge</td>
<td>gene expression vs gene expression</td>
</tr>
<tr class="even">
<td>gem</td>
<td>Gene expression of multiple genes</td>
</tr>
<tr class="odd">
<td>gec</td>
<td>Gene co-expression</td>
</tr>
<tr class="even">
<td>vio</td>
<td>Violin plot</td>
</tr>
<tr class="odd">
<td>pro</td>
<td>Proportion plot</td>
</tr>
<tr class="even">
<td>hea</td>
<td>Heatmap / Bubble / Dot plot</td>
</tr>
<tr class="odd">
<td>mar</td>
<td>Marker gene tables</td>
</tr>
<tr class="even">
<td>about</td>
<td>About page and additional info</td>
</tr>
</tbody>
</table>
<p>Of these, 8 tabs are displayed by default. Marker genes are not
displayed by default. More on that below. Specific tabs can be displayed
by setting the <code>tabs</code> argument. Tabs are displayed in the
order in which they are provided.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_conf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_config.html">create_config</a></span><span class="op">(</span><span class="va">pbmc_small</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/make_app.html">make_app</a></span><span class="op">(</span><span class="va">pbmc_small</span>, <span class="va">pbmc_conf</span>, gex.assay<span class="op">=</span><span class="st">"RNA"</span>, gex.slot<span class="op">=</span><span class="st">"data"</span>, </span>
<span>          shiny.title <span class="op">=</span> <span class="st">"PBMC"</span>, shiny.dir<span class="op">=</span><span class="st">"pbmc"</span>,</span>
<span>          tabs<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"civge"</span>,<span class="st">"gem"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">shiny</span><span class="fu">::</span><span class="fu"><a href="https://shiny.rstudio.com/reference/shiny/latest/runApp.html" class="external-link">runApp</a></span><span class="op">(</span><span class="st">"pbmc"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="images/selective-tabs.jpg" alt=""><p class="caption">Only displaying civge and gem tabs.</p>
</div>
</div>
<div class="section level3">
<h3 id="about-page">About page<a class="anchor" aria-label="anchor" href="#about-page"></a>
</h3>
<p>The <strong>about</strong> page is a tab that is displayed by
default. It contains acknowledgements by default and the user can modify
it as needed. Markdown formatting is supported. The
<strong>about</strong> page can be disabled by removing it from
<code>tabs</code>. When multiple datasets are used, a common
<strong>about</strong> page is created.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># app without about page</span></span>
<span><span class="va">pbmc_conf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_config.html">create_config</a></span><span class="op">(</span><span class="va">pbmc_small</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/make_app.html">make_app</a></span><span class="op">(</span><span class="va">pbmc_small</span>, <span class="va">pbmc_conf</span>, gex.assay<span class="op">=</span><span class="st">"RNA"</span>, gex.slot<span class="op">=</span><span class="st">"data"</span>, shiny.title <span class="op">=</span> <span class="st">"PBMC"</span>, shiny.dir<span class="op">=</span><span class="st">"app"</span>, tabs<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hea"</span>,<span class="st">"vio"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">shiny</span><span class="fu">::</span><span class="fu"><a href="https://shiny.rstudio.com/reference/shiny/latest/runApp.html" class="external-link">runApp</a></span><span class="op">(</span><span class="st">"app"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="marker-genes">Marker genes<a class="anchor" aria-label="anchor" href="#marker-genes"></a>
</h3>
<p>Marker gene functionality is only implemented for Seurat objects for
now. To display marker genes, <code><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">Seurat::FindAllMarkers()</a></code> need
to be run and resulting table added to the Seurat object slot
<code>misc</code> as a named list. To simplify this, one can use this
function:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#' @description Find all markers by identity</span></span>
<span><span class="co">#' @param obj Seurat v4 object</span></span>
<span><span class="co">#' @param ids Column name to used as identity (one or more)</span></span>
<span><span class="co">#' @param ... All arguments passed to Seurat::FindAllMarkers()</span></span>
<span><span class="co">#' </span></span>
<span><span class="va">fam</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">obj</span>,<span class="va">ids</span><span class="op">=</span><span class="cn">NULL</span>,<span class="va">...</span><span class="op">)</span><span class="op">{</span></span>
<span></span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/missing.html" class="external-link">missing</a></span><span class="op">(</span><span class="va">obj</span><span class="op">)</span><span class="op">)</span> <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="st">"Seurat object not specified."</span><span class="op">)</span></span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">ids</span><span class="op">)</span><span class="op">)</span> <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="st">"Argument 'ids' is NULL. It must be one or more column names from metadata."</span><span class="op">)</span></span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">is.character</a></span><span class="op">(</span><span class="va">ids</span><span class="op">)</span><span class="op">)</span> <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="st">"Argument 'ids' must be a character."</span><span class="op">)</span></span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="st">"Seurat"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">obj</span><span class="op">)</span><span class="op">)</span> <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="st">"Argument 'obj' must be a Seurat V4 object."</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/any.html" class="external-link">any</a></span><span class="op">(</span><span class="op">!</span><span class="va">ids</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">obj</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Following ids not found in metadata:"</span>,<span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">ids</span><span class="op">[</span><span class="op">!</span><span class="va">ids</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">obj</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span><span class="op">]</span>,collapse<span class="op">=</span><span class="st">","</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span></span>
<span>  <span class="va">fn</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">id</span><span class="op">=</span><span class="cn">NULL</span>,<span class="va">obj</span>,<span class="va">...</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">id</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>      <span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="va">id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">obj</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span><span class="op">)</span> <span class="kw"><a href="https://rdrr.io/r/base/stop.html" class="external-link">stop</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Following column does not exist: "</span>,<span class="va">id</span>,<span class="st">"."</span><span class="op">)</span><span class="op">)</span></span>
<span>      <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">obj</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">obj</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span>,<span class="va">id</span><span class="op">]</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">FindAllMarkers</a></span><span class="op">(</span><span class="va">obj</span>,<span class="va">...</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  </span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">ids</span>,<span class="va">fn</span>,obj<span class="op">=</span><span class="va">obj</span>,<span class="va">...</span><span class="op">)</span>,<span class="va">ids</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>One can compute marker genes for one or more metadata columns and
resulting named list of marker tables are stored in
<code>seurat@misc$markers</code>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># find marker genes for all levels in metadata column "letter.idents"</span></span>
<span><span class="va">pbmc_small</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu">fam</span><span class="op">(</span><span class="va">pbmc_small</span>,<span class="st">"letter.idents"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc_conf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_config.html">create_config</a></span><span class="op">(</span><span class="va">pbmc_small</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/make_app.html">make_app</a></span><span class="op">(</span><span class="va">pbmc_small</span>, <span class="va">pbmc_conf</span>, gex.assay<span class="op">=</span><span class="st">"RNA"</span>, gex.slot<span class="op">=</span><span class="st">"data"</span>, </span>
<span>          shiny.title <span class="op">=</span> <span class="st">"PBMC"</span>, shiny.dir<span class="op">=</span><span class="st">"pbmc"</span>,</span>
<span>          tabs<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mar"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">shiny</span><span class="fu">::</span><span class="fu"><a href="https://shiny.rstudio.com/reference/shiny/latest/runApp.html" class="external-link">runApp</a></span><span class="op">(</span><span class="st">"pbmc"</span><span class="op">)</span></span></code></pre></div>
<p><img src="images/mar.jpg"></p>
</div>
<div class="section level3">
<h3 id="theme">Theme<a class="anchor" aria-label="anchor" href="#theme"></a>
</h3>
<p>Change the app theme to another <a href="https://bootswatch.com/" class="external-link">bootswatch theme</a>. by changing the
argument <code>theme</code>. For more finer control over styling, edit
the <strong>www/styles.css</strong> file after creating the app.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_conf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_config.html">create_config</a></span><span class="op">(</span><span class="va">pbmc_small</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/make_app.html">make_app</a></span><span class="op">(</span><span class="va">pbmc_small</span>, <span class="va">pbmc_conf</span>, gex.assay<span class="op">=</span><span class="st">"RNA"</span>, gex.slot<span class="op">=</span><span class="st">"data"</span>, </span>
<span>          shiny.title <span class="op">=</span> <span class="st">"PBMC"</span>, shiny.dir<span class="op">=</span><span class="st">"pbmc"</span>, theme<span class="op">=</span><span class="st">"united"</span><span class="op">)</span></span>
<span><span class="fu">shiny</span><span class="fu">::</span><span class="fu"><a href="https://shiny.rstudio.com/reference/shiny/latest/runApp.html" class="external-link">runApp</a></span><span class="op">(</span><span class="st">"pbmc"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="images/theme-united.jpg" alt=""><p class="caption">Theme changed to ‘united’.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="session">Session<a class="anchor" aria-label="anchor" href="#session"></a>
</h2>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre class="foldable"><code><span><span class="co">## R version 4.2.2 (2022-10-31)</span></span>
<span><span class="co">## Platform: x86_64-conda-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Zorin OS 16.2</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS/LAPACK: /home/roy/miniconda3/envs/r-4.2/lib/libopenblasp-r0.3.21.so</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_GB.UTF-8        LC_COLLATE=en_GB.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] SeuratObject_4.1.3 Seurat_4.3.0       easyshiny_0.2      shiny_1.7.4       </span></span>
<span><span class="co">## [5] ggplot2_3.4.1      reticulate_1.28    hdf5r_1.3.8        data.table_1.14.8 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] Rtsne_0.16             colorspace_2.1-0       deldir_1.0-6          </span></span>
<span><span class="co">##   [4] ellipsis_0.3.2         ggridges_0.5.4         rprojroot_2.0.3       </span></span>
<span><span class="co">##   [7] fs_1.6.1               spatstat.data_3.0-0    leiden_0.4.3          </span></span>
<span><span class="co">##  [10] listenv_0.9.0          ggrepel_0.9.3          bit64_4.0.5           </span></span>
<span><span class="co">##  [13] fansi_1.0.4            codetools_0.2-19       splines_4.2.2         </span></span>
<span><span class="co">##  [16] cachem_1.0.6           knitr_1.42             polyclip_1.10-4       </span></span>
<span><span class="co">##  [19] jsonlite_1.8.4         ica_1.0-3              cluster_2.1.4         </span></span>
<span><span class="co">##  [22] png_0.1-8              uwot_0.1.14            spatstat.sparse_3.0-0 </span></span>
<span><span class="co">##  [25] sctransform_0.3.5      readr_2.1.4            compiler_4.2.2        </span></span>
<span><span class="co">##  [28] httr_1.4.4             Matrix_1.5-3           fastmap_1.1.0         </span></span>
<span><span class="co">##  [31] lazyeval_0.2.2         cli_3.6.0              later_1.3.0           </span></span>
<span><span class="co">##  [34] htmltools_0.5.4        tools_4.2.2            igraph_1.4.0          </span></span>
<span><span class="co">##  [37] gtable_0.3.1           glue_1.6.2             reshape2_1.4.4        </span></span>
<span><span class="co">##  [40] RANN_2.6.1             dplyr_1.1.0            Rcpp_1.0.10           </span></span>
<span><span class="co">##  [43] scattermore_0.8        jquerylib_0.1.4        pkgdown_2.0.7         </span></span>
<span><span class="co">##  [46] vctrs_0.5.2            nlme_3.1-162           spatstat.explore_3.0-6</span></span>
<span><span class="co">##  [49] progressr_0.13.0       lmtest_0.9-40          spatstat.random_3.1-3 </span></span>
<span><span class="co">##  [52] xfun_0.37              stringr_1.5.0          globals_0.16.2        </span></span>
<span><span class="co">##  [55] mime_0.12              miniUI_0.1.1.1         lifecycle_1.0.3       </span></span>
<span><span class="co">##  [58] irlba_2.3.5.1          goftest_1.2-3          future_1.31.0         </span></span>
<span><span class="co">##  [61] MASS_7.3-58.2          zoo_1.8-11             scales_1.2.1          </span></span>
<span><span class="co">##  [64] spatstat.utils_3.0-1   ragg_1.2.5             hms_1.1.2             </span></span>
<span><span class="co">##  [67] promises_1.2.0.1       parallel_4.2.2         RColorBrewer_1.1-3    </span></span>
<span><span class="co">##  [70] yaml_2.3.7             gridExtra_2.3          memoise_2.0.1         </span></span>
<span><span class="co">##  [73] pbapply_1.7-0          yulab.utils_0.0.6      sass_0.4.5            </span></span>
<span><span class="co">##  [76] stringi_1.7.12         desc_1.4.2             rlang_1.0.6           </span></span>
<span><span class="co">##  [79] pkgconfig_2.0.3        systemfonts_1.0.4      matrixStats_0.63.0    </span></span>
<span><span class="co">##  [82] evaluate_0.20          lattice_0.20-45        tensor_1.5            </span></span>
<span><span class="co">##  [85] ROCR_1.0-11            purrr_1.0.1            patchwork_1.1.2       </span></span>
<span><span class="co">##  [88] htmlwidgets_1.6.1      cowplot_1.1.1          bit_4.0.5             </span></span>
<span><span class="co">##  [91] tidyselect_1.2.0       parallelly_1.34.0      RcppAnnoy_0.0.20      </span></span>
<span><span class="co">##  [94] plyr_1.8.8             magrittr_2.0.3         R6_2.5.1              </span></span>
<span><span class="co">##  [97] generics_0.1.3         pillar_1.8.1           withr_2.5.0           </span></span>
<span><span class="co">## [100] fitdistrplus_1.1-8     abind_1.4-5            survival_3.5-3        </span></span>
<span><span class="co">## [103] sp_1.6-0               tibble_3.1.8           future.apply_1.10.0   </span></span>
<span><span class="co">## [106] KernSmooth_2.23-20     utf8_1.2.3             spatstat.geom_3.0-6   </span></span>
<span><span class="co">## [109] plotly_4.10.1          tzdb_0.3.0             rmarkdown_2.20        </span></span>
<span><span class="co">## [112] grid_4.2.2             digest_0.6.31          xtable_1.8-4          </span></span>
<span><span class="co">## [115] tidyr_1.3.0            httpuv_1.6.9           gridGraphics_0.5-1    </span></span>
<span><span class="co">## [118] textshaping_0.3.6      munsell_0.5.0          viridisLite_0.4.1     </span></span>
<span><span class="co">## [121] ggplotify_0.1.0        bslib_0.4.2</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://royfrancis.com" class="external-link">Roy Francis</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
