<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Generate data files required for shiny app — make_file • easyshiny</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Generate data files required for shiny app — make_file"><meta property="og:description" content="Generate data files required for shiny app. Five files will be generated, 
namely (i) the shinycell config prefix_conf.rds, (ii) the gene 
mapping object config prefix_gene.rds, (iii) the single-cell gene 
expression prefix_gexpr.h5, (iv) the single-cell metadata 
prefix_meta.rds and (v) the defaults for the Shiny app 
prefix_def.rds. A prefix is specified in each file to allow for the 
use of multiple single-cell datasets in one Shiny app. Note that both 
make_file and make_code functions are ran when 
running the wrapper function make_app."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">easyshiny</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/building.html">Building apps</a>
    </li>
    <li>
      <a href="../articles/deploy.html">Deploying your app</a>
    </li>
    <li>
      <a href="../articles/performance.html">Resource usage</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Versions</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/nbisweden/easyshiny">
    <span class="fa fa-github fa-lg fas"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Generate data files required for shiny app</h1>
    <small class="dont-index">Source: <a href="https://github.com/nbisweden/easyshiny/blob/HEAD/R/make_file.R"><code>R/make_file.R</code></a></small>
    <div class="hidden name"><code>make_file.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Generate data files required for shiny app. Five files will be generated, 
namely (i) the shinycell config <code>prefix_conf.rds</code>, (ii) the gene 
mapping object config <code>prefix_gene.rds</code>, (iii) the single-cell gene 
expression <code>prefix_gexpr.h5</code>, (iv) the single-cell metadata 
<code>prefix_meta.rds</code> and (v) the defaults for the Shiny app 
<code>prefix_def.rds</code>. A prefix is specified in each file to allow for the 
use of multiple single-cell datasets in one Shiny app. Note that both 
<code>make_file</code> and <code>make_code</code> functions are ran when 
running the wrapper function <code>make_app</code>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">make_file</span><span class="op">(</span></span>
<span>  <span class="va">obj</span>,</span>
<span>  <span class="va">scConf</span>,</span>
<span>  gex.assay <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  gex.slot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="st">"scale.data"</span>, <span class="st">"counts"</span><span class="op">)</span>,</span>
<span>  gene.mapping <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  shiny.prefix <span class="op">=</span> <span class="st">"sc1"</span>,</span>
<span>  shiny.dir <span class="op">=</span> <span class="st">"shinyApp/"</span>,</span>
<span>  default.gene1 <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  default.gene2 <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  default.multigene <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  default.dimred <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  chunkSize <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  mar <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>obj</dt>
<dd><p>input single-cell object for Seurat (v3+) / SingleCellExperiment 
data or input file path for h5ad / loom files</p></dd>


<dt>scConf</dt>
<dd><p>shinycell config data.table</p></dd>


<dt>gex.assay</dt>
<dd><p>assay in single-cell data object to use for plotting 
gene expression, which must match one of the following:</p><ul><li><p>Seurat objects: "RNA" or "integrated" assay, 
    default is "RNA"</p></li>
<li><p>SCE objects: "logcounts" or "normcounts" or "counts", 
    default is "logcounts"</p></li>
<li><p>h5ad files: "X" or any assay in "layers",
    default is "X"</p></li>
<li><p>loom files: "matrix" or any assay in "layers",
    default is "matrix"</p></li>
</ul></dd>


<dt>gex.slot</dt>
<dd><p>slot in single-cell assay to plot. This is only used 
for Seurat objects (v3+). Default is to use the "data" slot</p></dd>


<dt>gene.mapping</dt>
<dd><p>specifies whether to convert human / mouse Ensembl gene 
IDs (e.g. ENSG000xxx / ENSMUSG000xxx) into "user-friendly" gene symbols. 
Set this to <code>TRUE</code> if you are using Ensembl gene IDs. Default is 
<code>FALSE</code> which is not to perform any conversion. Alternatively, users 
can supply a named vector where <code>names(gene.mapping)</code> correspond 
to the actual gene identifiers in the gene expression matrix and 
<code>gene.mapping</code> correspond to new identifiers to map to</p></dd>


<dt>shiny.prefix</dt>
<dd><p>specify file prefix</p></dd>


<dt>shiny.dir</dt>
<dd><p>specify directory to create the shiny app in</p></dd>


<dt>default.gene1</dt>
<dd><p>specify primary default gene to show</p></dd>


<dt>default.gene2</dt>
<dd><p>specify secondary default gene to show</p></dd>


<dt>default.multigene</dt>
<dd><p>character vector specifying default genes to 
show in bubbleplot / heatmap</p></dd>


<dt>default.dimred</dt>
<dd><p>character vector specifying the two default dimension 
reductions. Default is to use UMAP if not TSNE embeddings</p></dd>


<dt>chunkSize</dt>
<dd><p>number of genes written to h5file at any one time. Lower 
this number to reduce memory consumption. Should not be less than 10</p></dd>


<dt>mar</dt>
<dd><p>Logical indicating if marker genes are used</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>data files required for shiny app</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>John F. Ouyang</p>
<p>Roy Francis</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://royfrancis.com" class="external-link">Roy Francis</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

